ga4gh/avrotools.py

```diff
@@ -0,0 +1,83 @@
+"""
+Provides additional functionality based on avro
+"""
+from __future__ import division
+from __future__ import print_function
+from __future__ import unicode_literals
+
+import json
+import avro.io
+
+import ga4gh.protocol as protocol
+
+
+class ValidationToolException(Exception):
+    """
+    Something that went wrong in the ValidationTool
+    """
+
+
+class ValidationTool(object):
+    """
+    Provides additional validation methods based on avro
+    """
+    @classmethod
+    def getInvalidFields(cls, class_, jsonDict):
+        """
+        Takes a ProtocolElement subclass class_ and a jsonDict instantiation
+        of that class.  Returns a dictionary whose keys are the names of
+        fields in the jsonDict that failed validation.  The values of the
+        returned dictionary are the jsonDict's values for those fields.
+
+        This is an expensive function to run and shouldn't be on the main
+        data path.
+
+        TODO is there an easier/faster way to do this than generating
+        an on-demand schema for every field we want to test?
+        """
+        if not issubclass(class_, protocol.ProtocolElement):
+            raise ValidationToolException(
+                "class must be subclass of ProtocolElement")
+        invalidFields = {}
+        schema = class_.schema
+        schemaJson = schema.to_json()
+        for field in schemaJson['fields']:
+            fieldName = field['name']
+            fieldValue = jsonDict[fieldName]
+            fieldSchema = cls._createFieldSchema(schema, fieldName)
+            fieldDict = {fieldName: fieldValue}
+            isFieldValid = avro.io.validate(fieldSchema, fieldDict)
+            if not isFieldValid:
+                invalidFields[fieldName] = fieldValue
+        return invalidFields
+
+    @classmethod
+    def _createFieldSchema(cls, schema, fieldName):
+        schemaJson = schema.to_json()
+        # TODO this is an ugly hack
+        # I think GAReadAlignment.nextMatePosition's schema is not getting
+        # written correctly.
+        # In the schema its type attribute is:
```

Comment:
Sure.  This is a schemas issue so I don't think it should hold up getting this change in.  We can work around it for the time being.